File: 64. Minimum Path Sum/64. Minimum Path Sum bottom-up.py
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        dp = [[0]*n for _ in range(m)]

        for row in range(m):
            for column in range(n):
                dp[row][column] += grid[row][column]
                if row > 0 and column > 0:
                    dp[row][column] += min(dp[row-1]
                                           [column], dp[row][column-1])
                elif row > 0:
                    dp[row][column] += dp[row-1][column]
                elif column > 0:
                    dp[row][column] += dp[row][column-1]
        return dp[m-1][n-1]

File: 64. Minimum Path Sum/64. Minimum Path Sum bottom-up.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dp = new int[m][n];

        for (int row = 0; row < m; row++) {
            for (int column = 0; column < n; column++) {
                dp[row][column] += grid[row][column];
                if (row > 0 && column > 0) {
                    dp[row][column] += Math.min(dp[row - 1][column], dp[row][column - 1]);
                } else if (row > 0) {
                    dp[row][column] += dp[row - 1][column];
                } else if (column > 0) {
                    dp[row][column] += dp[row][column - 1];
                }
            }
        }
        return dp[m - 1][n - 1];
    }
}